app/src/main/java/nodomain/freeyourgadget/gadgetbridge/export/GPXExporter.java

```diff
@@ -87,7 +88,7 @@ private void exportTrack(XmlSerializer ser, ActivityTrack track) throws IOExcept
         List<ActivityPoint> trackPoints = track.getTrackPoints();
         String source = getSource(track);
         for (ActivityPoint point : trackPoints) {
-            exportTrackPoint(ser, point, source);
+            exportTrackPoint(ser, point, source, trackPoints);
```

Comment:
AFAICS this implements some kind of nearest-neighbour algorithm to find a valid heart rate value for those track points that lack one.

I would like to see this optimized a bit, since currently we would traverse the entire list of trackpoints again and again and again.

A simple optimization would be to assume that the samples are sorted by increasing timestamp. After finding the hr-sample for the first sample without hr, you would remember the index of the hr-sample and for the next sample you would start from there instead of from the beginning.

Additionally you could perform a custom binary search by the timestamp.

And last, I think there should be a maximum distance. If the nearest neighbour hr-value is 1h away, there's no point in using it. I think the maximum distance should be 1min or so.

Reply:
Hi, If possible please do the optimization, I am purly c#/.net developer and for me is quite difficult to write code in java syntax :) not mention about android studio :) that why my code is a mess :)